build(cmake): Add support for building library targets that contain source files and linked libraries.

Bill-hbrhbr opened the pull request:
<!-- markdownlint-disable MD012 -->

<!--
Set the PR title to a meaningful commit message that:

* is in imperative form.
* follows the Conventional Commits specification (https://www.conventionalcommits.org).
  * See https://github.com/commitizen/conventional-commit-types/blob/master/index.json for possible
    types.

Example:

fix: Don't add implicit wildcards ('*') at the beginning and the end of a query (fixes #390).
-->

# Description

<!-- Describe what this request will change/fix and provide any details necessary for reviewers. -->

Expanded and updated the arglist of [`cpp_library()`](https://github.com/y-scope/ystdlib-cpp/pull/17):

```cmake
# @param NAME
# @param NAMESPACE
# @param PUBLIC_HEADERS
# @param PRIVATE_SOURCES
# @param PUBLIC_LINK_LIBRARIES
# @param PRIVATE_LINK_LIBRARIES
# @parms TESTS_SOURCES
# @param [BUILD_INCLUDE_DIR="${PROJECT_SOURCE_DIR}/src"]
```

Added a helper function `check_if_header_only()` for two functionalities:
* Enforce that there are only header files inside `PUBLIC_HEADERS`
* Check whether there are header files inside `PRIVATE_SOURCES`
  * If yes, build a `static/shared` library target
  * If no, build an `interface` library target


# Checklist

<!-- Ensure each item below is satisfied and indicate so by inserting an `x` within each `[ ]`. -->

* [X] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [X] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [X] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

<!-- Describe what tests and validation you performed on the change. -->

No tests needed.

[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced new build options allowing users to choose shared library builds and to enable testing.
  - Enhanced library configuration to better distinguish header-only from full implementation libraries with improved error notifications.
  - Refined the public interfaces for container and error-handling modules to ensure a clearer and more robust integration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Bill-hbrhbr commented:
@coderabbitai
CMake/ystdlib-cpp-helpers.cmake (2)
1-18: Well-structured header-only detection logic
currently we only lint h, hpp, cpp files. no need to add more to the scope

davidlion commented on CMake/ystdlib-cpp-helpers.cmake:
```diff
@@ -16,6 +39,10 @@ function(cpp_library)
     set(multiValueArgs
+        HEADERS
+        SOURCES
+        DEPENDS
+        INTERNAL_DEPENDS
```
Lets match the cmake naming so something like:
`PUBLIC_LINK_LIBS` (and `PRIVATE...`) or just `PUBLIC_LIBS` maybe.

Bill-hbrhbr replied:
I've renamed everything to use very explicit names. I got the idea of `INTERFACE_LIBRARIES` from [here](https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_LINK_LIBRARIES.html)

davidlion commented on CMake/ystdlib-cpp-helpers.cmake:
```diff
@@ -16,6 +39,10 @@ function(cpp_library)
         NAMESPACE
     )
     set(multiValueArgs
+        HEADERS
+        SOURCES
```
I think everything works even if we don't separate/distinguish headers and source files, so I think just `SOURCES` is fine. If it is important to separate them, then maybe we should concatenate them and pass both to `check_if_header_only_library` to avoid user error (currently if you put a cpp file in `HEADERS` I think things will break)?

Bill-hbrhbr replied:
On a second thought, I've decided to keep them split and renamed the params for more clarity. I've also enforced a check of the public header list.

Bill-hbrhbr commented:
TLDR of changes since last review (except for the ones directly resolving your comments):
1. Renamed param names of `cpp_library()` and made them more verbose for more clarity.
2. Added a check to enforce header-only files in `INTERFACE_HEADERS` by making a small update to the helper function
3. Added a variable that stores the alias target name and updated all reference sites.

davidlion commented on CMake/ystdlib-cpp-helpers.cmake:
```diff
@@ -25,74 +22,81 @@ endfunction()
     else()
-        # Empty string makes library type determined by `BUILD_SHARED_LIBS` option
-        add_library(${arg_cpp_lib_NAME} "")
+        # Library type defaults to static if `BUILD_SHARED_LIBS` is OFF and to shared if ON.
```
This statement isn't really correct. If BUILD_SHARED_LIBS is set then there is no default value used, the set value is used. We want to document the behaviour when it is not set (aka what CMake will do by default).

davidlion commented on CMake/ystdlib-cpp-helpers.cmake:
```diff
@@ -22,12 +22,12 @@ endfunction()
-# @param [INTERFACE_BUILD_INCLUDE_DIR="${PROJECT_SOURCE_DIR}/src"] The list of include paths for
+# @param [BUILD_INCLUDE_DIR="${PROJECT_SOURCE_DIR}/src"] The list of include paths for
 # building the library and for external projects that builds `ystdlib-cpp` as a CMAKE subproject
 # via the add_subdirectory() function.
```
After changing the variable name the text needs to be re-flowed.
Unfortunately, I can't make the suggestion since the other lines aren't in the diff.
